File: frontend/src/types/sre.ts
export interface SliMetricsItem {
  total_requests: number;
  failed_requests: number;
  availability: number;
  error_rate: number;
  latency_p50_ms: number;
  latency_p95_ms: number;
  latency_p99_ms: number;
  throughput_rps: number;
}

export interface SloItem {
  id: string;
  service: string;
  name: string;
  description?: string | null;
  indicator_type: "availability" | "latency" | "error_rate" | "throughput" | string;
  target: number;
  target_threshold_ms?: number | null;
  window: string;
  enabled: bool;
  current_sli: number;
  compliance_percentage: number;
  status: "HEALTHY" | "AT_RISK" | "BREACHED" | string;
  created_at: string;
  updated_at: string;
}

export interface SloCreatePayload {
  service: string;
  name: string;
  description?: string;
  indicator_type?: string;
  target: number;
  target_threshold_ms?: number;
  window?: string;
  enabled?: boolean;
}

export interface SloListResponse {
  slos: SloItem[];
  total: number;
}

export interface ErrorBudgetItem {
  service: string;
  target_slo: number;
  total_budget_pct: number;
  consumed_pct: number;
  remaining_pct: number;
  remaining_budget_units: number;
  status: "HEALTHY" | "AT_RISK" | "EXHAUSTED" | string;
}

export interface BurnRateItem {
  service: string;
  burn_1h: number;
  burn_6h: number;
  burn_24h: number;
  burn_7d: number;
  status: "NORMAL" | "ELEVATED" | "CRITICAL" | string;
}

export interface ServiceReliabilityItem {
  service: string;
  reliability_score: number;
  rating: "EXCELLENT" | "GOOD" | "DEGRADED" | "CRITICAL" | string;
  availability: number;
  latency_p95_ms: number;
  error_rate: number;
  throughput_rps: number;
  slo_status: "HEALTHY" | "AT_RISK" | "BREACHED" | string;
  error_budget_remaining_pct: number;
  burn_rate_status: "NORMAL" | "ELEVATED" | "CRITICAL" | string;
  active_incidents_count: number;
  trend: "IMPROVING" | "STABLE" | "DEGRADED" | string;
}

export interface ServiceReliabilityListResponse {
  services: ServiceReliabilityItem[];
  total: number;
}

export interface ReliabilityRiskItem {
  id: string;
  risk: string;
  severity: "CRITICAL" | "HIGH" | "MEDIUM" | "LOW" | string;
  service: string;
  metric: string;
  current_value: string;
  threshold: string;
  detected_at: string;
  explanation: string;
  recommended_action: string;
}

export interface ReliabilityRiskListResponse {
  risks: ReliabilityRiskItem[];
  total_risks: number;
  critical_risks: number;
}

export interface IncidentImpactItem {
  id: string;
  title: string;
  service: string;
  severity: string;
  status: string;
  started_at?: string | null;
  duration_minutes: number;
  slo_impact: string;
  budget_impact_pct: number;
}

export interface IncidentImpactListResponse {
  incidents: IncidentImpactItem[];
  total: number;
}

export interface DependencyImpactItem {
  dependency: string;
  target_service: string;
  health: "HEALTHY" | "DEGRADED" | "CRITICAL" | string;
  latency_ms: number;
  error_rate: number;
  affected_services: string[];
  reliability_risk: string;
}

export interface DependencyImpactListResponse {
  dependencies: DependencyImpactItem[];
  total: number;
}

export interface ForecastPeriod {
  availability: number;
  error_rate: number;
  latency_ms: number;
  slo_status: string;
}

export interface ReliabilityForecastResponse {
  forecast_24h: ForecastPeriod;
  forecast_7d: ForecastPeriod;
  forecast_30d: ForecastPeriod;
  confidence: number;
  historical_basis: string;
  status: string;
}

export interface SreRecommendationItem {
  id: string;
  service: string;
  category: string;
  severity: "CRITICAL" | "HIGH" | "MEDIUM" | "LOW" | string;
  reason: string;
  evidence: string;
  recommended_action: string;
  expected_impact: string;
  confidence: number;
}

export interface SreRecommendationListResponse {
  recommendations: SreRecommendationItem[];
  total: number;
}

export interface SreOverviewResponse {
  overall_score: number;
  overall_rating: string;
  services_healthy: number;
  services_at_risk: number;
  slo_breaches: number;
  error_budget_remaining_avg: number;
  active_incidents_count: number;
  data_source?: string;
  environment?: string;
}

export interface SreAnalyzeResponse {
  executive_summary: string;
  critical_services: string[];
  error_budget_warnings: string[];
  sre_recommendations: SreRecommendationItem[];
  analyzed_at: string;
  analysis_engine?: string;
}

File: backend/app/schemas/sre.py
"""
Pydantic schemas for Enterprise SRE & Reliability Intelligence Center.
"""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, Field

# ── SLI & SLO Schemas ─────────────────────────────────────────────────────────


class SliMetricsItem(BaseModel):
  total_requests: int
  failed_requests: int
  availability: float
  error_rate: float
  latency_p50_ms: float
  latency_p95_ms: float
  latency_p99_ms: float
  throughput_rps: float


class SloItem(BaseModel):
  id: uuid.UUID
  service: str
  name: str
  description: str | None = None
  indicator_type: str  # availability | latency | error_rate | throughput
  target: float
  target_threshold_ms: float | None = None
  window: str  # 30d | 7d | 24h | 1h
  enabled: bool
  current_sli: float = 0.0
  compliance_percentage: float = 100.0
  status: str  # HEALTHY | AT_RISK | BREACHED
  created_at: datetime
  updated_at: datetime

  model_config = {"from_attributes": True}


class SloCreatePayload(BaseModel):
  service: str = Field(..., min_length=2)
  name: str = Field(..., min_length=3)
  description: str | None = None
  indicator_type: str = Field(default="availability")
  target: float = Field(..., gt=0.0, le=100.0)
  target_threshold_ms: float | None = None
  window: str = Field(default="30d")
  enabled: bool = Field(default=True)


class SloUpdatePayload(BaseModel):
  name: str | None = None
  description: str | None = None
  target: float | None = None
  target_threshold_ms: float | None = None
  window: str | None = None
  enabled: bool | None = None


class SloListResponse(BaseModel):
  slos: list[SloItem]
  total: int


# ── Error Budget & Burn Rate Schemas ──────────────────────────────────────────


class ErrorBudgetItem(BaseModel):
  service: str
  target_slo: float
  total_budget_pct: float
  consumed_pct: float
  remaining_pct: float
  remaining_budget_units: float
  status: str  # HEALTHY | AT_RISK | EXHAUSTED


class BurnRateItem(BaseModel):
  service: str
  burn_1h: float
  burn_6h: float
  burn_24h: float
  burn_7d: float
  status: str  # NORMAL | ELEVATED | CRITICAL


# ── Service Reliability Item & List ───────────────────────────────────────────


class ServiceReliabilityItem(BaseModel):
  service: str
  reliability_score: float
  rating: str  # EXCELLENT | GOOD | DEGRADED | CRITICAL
  availability: float
  latency_p95_ms: float
  error_rate: float
  throughput_rps: float
  slo_status: str  # HEALTHY | AT_RISK | BREACHED
  error_budget_remaining_pct: float
  burn_rate_status: str  # NORMAL | ELEVATED | CRITICAL
  active_incidents_count: int
  trend: str  # IMPROVING | STABLE | DEGRADED


class ServiceReliabilityListResponse(BaseModel):
  services: list[ServiceReliabilityItem]
  total: int


# ── Risk, Incident & Dependency Schemas ───────────────────────────────────────


class ReliabilityRiskItem(BaseModel):
  id: str
  risk: str
  severity: str  # CRITICAL | HIGH | MEDIUM | LOW
  service: str
  metric: str
  current_value: str
  threshold: str
  detected_at: str
  explanation: str
  recommended_action: str


class ReliabilityRiskListResponse(BaseModel):
  risks: list[ReliabilityRiskItem]
  total_risks: int
  critical_risks: int


class IncidentImpactItem(BaseModel):
  id: uuid.UUID
  title: str
  service: str
  severity: str
  status: str
  started_at: datetime | None = None
  duration_minutes: float = 0.0
  slo_impact: str
  budget_impact_pct: float


class IncidentImpactListResponse(BaseModel):
  incidents: list[IncidentImpactItem]
  total: int


class DependencyImpactItem(BaseModel):
  dependency: str
  target_service: str
  health: str  # HEALTHY | DEGRADED | CRITICAL
  latency_ms: float
  error_rate: float
  affected_services: list[str]
  reliability_risk: str


class DependencyImpactListResponse(BaseModel):
  dependencies: list[DependencyImpactItem]
  total: int


# ── Forecast & Recommendation Schemas ─────────────────────────────────────────


class ForecastPeriod(BaseModel):
  availability: float
  error_rate: float
  latency_ms: float
  slo_status: str


class ReliabilityForecastResponse(BaseModel):
  forecast_24h: ForecastPeriod
  forecast_7d: ForecastPeriod
  forecast_30d: ForecastPeriod
  confidence: float
  historical_basis: str
  status: str  # VALID | INSUFFICIENT_DATA


class SreRecommendationItem(BaseModel):
  id: str
  service: str
  category: str
  severity: str  # CRITICAL | HIGH | MEDIUM | LOW
  reason: str
  evidence: str
  recommended_action: str
  expected_impact: str
  confidence: float


class SreRecommendationListResponse(BaseModel):
  recommendations: list[SreRecommendationItem]
  total: int


# ── SRE Overview & AI Analysis Schemas ────────────────────────────────────────


class SreOverviewResponse(BaseModel):
  overall_score: float
  overall_rating: str
  services_healthy: int
  services_at_risk: int
  slo_breaches: int
  error_budget_remaining_avg: float
  active_incidents_count: int
  data_source: str = Field(
      default="Demo Data — No Production Telemetry Connected"
  )
  environment: str = Field(default="Local Development")


class SreAnalyzeResponse(BaseModel):
  executive_summary: str
  critical_services: list[str]
  error_budget_warnings: list[str]
  sre_recommendations: list[SreRecommendationItem]
  analyzed_at: str
  analysis_engine: str = Field(default="Local SRE Intelligence")
